packages/stable/src/api/documents/types.ts

```diff
@@ -0,0 +1,369 @@
+    }
+  | {
+      geojsonWithin: DocumentFilterPropertyAndGeometry;
+    };
+
+type DocumentFilterProperty = { property: string[] };
+
+type DocumentFilterValue = { value: string | number | boolean | DocumentLabel };
+
+type DocumentFilterValues = {
+  values: string[] | number[] | boolean[] | DocumentLabel[];
+};
+
+type DocumentFilterRangeValues = {
+  gt?: number;
+  lt?: number;
+  gte?: number;
+  lte?: number;
+};
+
+type DocumentFilterGeometry = { geometry: DocumentGeometry };
+
+type DocumentFilterPropertyAndValue = DocumentFilterProperty &
+  DocumentFilterValue;
+
+type DocumentFilterPropertyAndValues = DocumentFilterProperty &
+  DocumentFilterValues;
+
+type DocumentFilterPropertyAndRangeValues = DocumentFilterProperty &
+  DocumentFilterRangeValues;
+
+type DocumentFilterPropertyAndGeometry = DocumentFilterProperty &
+  DocumentFilterGeometry;
+
+export interface DocumentCountAggregate {
+  /** User defined name for this aggregate */
+  name: string;
+
+  /**
+   * count
+   * @example count
+   */
+  aggregate: string;
+
+  /** List of fields to group the count by. It is currently only possible to group by 1 field or 0 fields. If grouping by 0 fields, the aggregate value is the total count of all documents. */
+  groupBy?: DocumentFilterProperty[];
+}
+
+/**
+ * The number of milliseconds since 00:00:00 Thursday, 1 January 1970, Coordinated Universal Time (UTC), minus leap seconds.
+ * @format int64
+ * @min 0
+ * @example 1638795554528
+ */
+export type DocumentTimestamp = number;
+
+/**
+ * The source file that this document is derived from.
+ */
+export interface DocumentSourceFile {
+  /**
+   * Name of the file
+   * @example hamlet.txt
+   */
+  name: string;
+
+  /**
+   * The directory the file can be found in
+   * @example plays/shakespeare
+   */
+  directory?: string;
+
+  /**
+   * The source of the file
+   * @example SubsurfaceConnectors
+   */
+  source?: string;
+
+  /**
+   * The mime type of the file
+   * @example application/octet-stream
+   */
+  mimeType?: string;
+
+  /**
+   * The size of the source file in bytes
+   * @format int64
+   * @example 1000
+   */
+  size?: number;
+
+  /**
+   * The ids of the assets related to this file
+   * @example []
+   */
+  assetIds?: CogniteInternalId[];
+
+  /**
+   * A list of labels associated with this document's source file in CDF.
+   * @example [{"externalId":"play"},{"externalId":"tragedy"}]
+   */
+  labels?: DocumentLabel[];
+
+  /** GeoJson representation of a geometry. */
+  geoLocation?: DocumentGeometry;
+
+  /** The id if the dataset this file belongs to, if any */
+  datasetId?: CogniteInternalId;
+
+  /**
+   * The security category IDs required to access this file
+   * @example []
+   */
+  securityCategories?: number[];
+  metadata?: Record<string, string>;
+}
+
+/**
+ * GeoJson representation of a geometry.
+ */
+export interface DocumentGeometry {
+  /** Type of the shape. Currently we support "polygon", "linestring" and "point". */
```

Comment:
Are the types in this file hand-written or somehow generated? If the latter case it would be nice with some instructions for how to keep it up-to-date with future API changes.

I notice that the comments here is not the same as in the API docs, and the values here is no longer correct. It would also be nice if the type names matched the schema used for the API, to reduce cognitive load and reduce type name changes in future automation (as clients of this library will have to migrate their code if they import the types).

Reply:
It's hand-written. Well, I copied the playground types and modified them to match the v1 spec.

I suppose I could remove these comments. I would not like to manually have to keep them in sync with the spec. That sounds like a lot of work for little benefit.

Reply:
I think this is a strong argument for generating the types (#740).

We should focus on the developer experience, so comments should be included. However I'd rather exclude comments than including the wrong comments, if we have to choose.

Since we are not there yet (generating types), can we try to be compatible with it? To preserve developer experience we don't want to rename these later.

Since I spotted at least another defect in the types - maybe @andersfylling can help generating this for a one-shot now? I think this was initially generated for the playground API. That way we get proper names, fix other defects to the types, and will be compatible with future code gen as far as I understand without introducing breaking changes.

Reply:
The v1 spec is funny. I can generate the entire playground api types, without issues, but the v1 might have some weird cyclic reference. Causing a never ending lookup loop (suspicion). It keeps generating forever, ending up with a crash due to memory usage. Even 4GiB isn't enough.

Might be that the script just have to check the type record before exploring a type to avoid redundancy.

Reply:
The new filter specification is definitely cyclic. In particular, the bool filters refer back to the common filter type